src/tools/fuzzing.h

```diff
@@ -432,21 +461,83 @@ class TranslateToFuzzReader {
   void setupEvents() {
     Index num = upTo(3);
     for (size_t i = 0; i < num; i++) {
-      auto* event =
-        builder.makeEvent(std::string("event$") + std::to_string(i),
-                          WASM_EVENT_ATTRIBUTE_EXCEPTION,
-                          Signature(getControlFlowType(), Type::none));
-      wasm.addEvent(event);
+      auto name = std::string("event$") + std::to_string(i);
+      if (!wasm.getEventOrNull(name)) {
+        auto* event =
+          builder.makeEvent(name,
+                            WASM_EVENT_ATTRIBUTE_EXCEPTION,
+                            Signature(getControlFlowType(), Type::none));
+        wasm.addEvent(event);
+      }
+    }
+  }
+
+  void finalizeMemory() {
+    for (auto& segment : wasm.memory.segments) {
+      Address maxOffset = segment.data.size();
+      if (!segment.isPassive) {
+        // If the offset is a global that was imported (which is ok) but no
+        // longer is (not ok) we need to change that.
+        if (auto* offset = segment.offset->dynCast<GlobalGet>()) {
+          if (!wasm.getGlobal(offset->name)->imported()) {
+            // TODO: the segments must not overlap...
+            segment.offset =
+              builder.makeConst(Literal::makeFromInt32(0, Type::i32));
+          }
+        }
+        if (auto* offset = segment.offset->dynCast<Const>()) {
+          maxOffset = maxOffset + offset->value.getInteger();
+        }
+      }
+      wasm.memory.initial = std::max(
+        wasm.memory.initial,
+        Address((maxOffset + Memory::kPageSize - 1) / Memory::kPageSize));
+    }
+    wasm.memory.initial = std::max(wasm.memory.initial, USABLE_MEMORY);
+    // Avoid an unlimited memory size, which would make fuzzing very difficult
+    // as different VMs will run out of system memory in different ways.
+    if (wasm.memory.max == Memory::kUnlimitedSize) {
+      wasm.memory.max = wasm.memory.initial;
     }
+    if (wasm.memory.max <= wasm.memory.initial) {
+      // To allow growth to work (which a testcase may assume), try to make the
+      // maximum larger than the initial.
+      // TODO: scan the wasm for grow instructions?
+      wasm.memory.max =
+        std::min(Address(wasm.memory.initial + 1), Address(Memory::kMaxSize32));
+    }
+    // See above for globals.
```

Comment:
Does this mean we don't allow imported memory? It this is all I guess we can just say that here, because it's one-line anyway.. The same for the same comment in `finalizeTable`.

Reply:
Done.